airflow/www/views.py

```diff
@@ -778,42 +787,46 @@ def log(self):
-                    try:
-                        import requests
-                        timeout = None  # No timeout
+                for i in xrange(len(logs)):
+                    local_log_path = filename(local_log_dir, i)
+                    if os.path.exists(local_log_path):
                         try:
-                            timeout = conf.getint('webserver', 'log_fetch_timeout_sec')
-                        except (AirflowConfigException, ValueError):
-                            pass
-
-                        response = requests.get(url, timeout=timeout)
-                        response.raise_for_status()
-                        log += '\n' + response.text
-                    except:
-                        log += "*** Failed to fetch log file from worker.\n".format(
-                            **locals())
-
-        if PY2 and not isinstance(log, unicode):
-            log = log.decode('utf-8')
+                            f = open(local_log_path)
+                            logs[i] += "*** Reading local log.\n" + "".join(f.readlines())
+                            f.close()
+                        except:
+                            logs[i] = "*** Failed to load local log file: {0}.\n".format(local_log_path)
+                    else:
+                        worker_log_server_port = \
+                            conf.get('celery', 'WORKER_LOG_SERVER_PORT')
+                        log_relative_path = filename(log_relative_dir, i)
+                        url = os.path.join(
+                            "http://{ti.hostname}:{worker_log_server_port}/log", log_relative_path
+                        ).format(**locals())
+                        logs[i] += "*** Log file isn't local.\n"
+                        logs[i] += "*** Fetching here: {url}\n".format(**locals())
+                        try:
+                            import requests
+                            timeout = None  # No timeout
+                            try:
+                                timeout = conf.getint('webserver', 'log_fetch_timeout_sec')
+                            except (AirflowConfigException, ValueError):
+                                pass
+
+                            response = requests.get(url, timeout=timeout)
+                            response.raise_for_status()
+                            log += '\n' + response.text
+                        except:
+                            logs[i] += "*** Failed to fetch log file from worker.\n".format(
+                                **locals())
+
+        for i, log in enumerate(logs):
+            if PY2 and not isinstance(log, unicode):
+                logs[i] = log.decode('utf-8')
 
         return self.render(
-            'airflow/ti_code.html',
-            code=log, dag=dag, title="Log", task_id=task_id,
+            'airflow/ti_logs.html',
+            logs=logs[::-1], dag=dag, title="Log", task_id=task_id,
```

Comment:
Doesn't the page need to know which try # it is serving (assuming that the active tab is highlighted in the UI somehow).